### PR TITLE
DE-3599: Add parsing function stubs and tests

### PR DESCRIPTION
**Pull request by abclay**
## JIRA ticket
https://tetrascience.atlassian.net/browse/DE-3599

## Description

I have added all the test cases from [Confluence](https://tetrascience.atlassian.net/wiki/spaces/DE/pages/3230400808/Unit+Test+Cases+for+Parsing+Primitives#Parse-string-to-boolean)
I have also added stub functions to be tested.  I wrote very basic implementations to check my tests and left them in. When we do the next ticket to implement these methods, we can use the reference implementations as a baseline.

## Automated testing

- [x] Unit tests
- [ ] Integration tests
- [ ] API tests
- [ ] End-To-End tests
- [ ] N/A (please clarify)

### CODE REVIEW

**Comment by mlipschultz** [comment]
Before this merges, I'd like to get approval on the tests (either here or on the [Confluence page](https://tetrascience.atlassian.net/wiki/spaces/DE/pages/3230400808/Unit+Test+Cases+for+Parsing+Primitives)) from @wkopen, @evan-anderson-ts, @igridchyn, @tlitetrasci, or @robjhornby-ts.  Basically, I want to make sure that someone very familiar with these parsing needs approves our test cases before we start researching/implementing a solution to pass those test cases, otherwise we might waste time/effort on the wrong solution.

File: task_script_utils/parse.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+def to_boolean(
+    value: str, true_set: Set[str], false_set: Set[str], case_sensitive: bool = False
+) -> Optional[bool]:
+    """
+    Converts given string to a boolean value based on whether it is in either of the sets provided
+
+    :param value: A string to convert
+    :type value: string
+    :param true_set: A set of strings which represent True values
+    :type true_set: Set[string]
```
Comment (VladimirShteyn) [inline task_script_utils/parse.py:64]: IMO we shouldn't specify type in the docstring. Doing so creates two sources of truth: one in the function signature and another in the docstring. I recommend deleting `:type ...:` throughout.

File: __tests__/unit/parse_functions/test_to_int.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+@pytest.mark.parametrize(
+    "value",
+    [
+        "two",
+        "12.4.5",
+        "1/2",
+        "4 + 2i",
+        "5 + 3j",
+        "2,500",
+        "0xbadbeef",
```
Comment (VladimirShteyn) [inline __tests__/unit/parse_functions/test_to_int.py:231]: 🥩

nom!

File: __tests__/unit/parse_functions/test_to_int.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    value = "-2.74e-3"
+
+    # Act
+    actual = parse.to_int(value)
+
+    # Assert
+    assert actual is None
+
+
+def test_to_int_leading_whitespace():
```
Comment (VladimirShteyn) [inline __tests__/unit/parse_functions/test_to_int.py:146]: [minor] For readability, consider grouping the whitespace tests into a `parametrize` or a single class.

File: __tests__/unit/parse_functions/test_to_int.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    """Test Scientific Notation, negative exponent, between 0 and 1."""
+    # Arrange from parameters
+    # Act
+    actual = parse.to_int(value)
+
+    # Assert
+    assert actual is None
+
+
+def test_to_int_scientific_notation_negative_exponent_greater_than_1():
```
Comment (VladimirShteyn) [inline __tests__/unit/parse_functions/test_to_int.py:108]: I recommend adding the following test cases where the exponent is zero,  probably in a `parametrize`. That is,

```
"1e0", "1e+0", "1e-0",
"1.0e0", "1.0e+0", "1.0e-0",
"100e0", "100e+0", "100e-0",
"100.0e0", "100.0e+0", "100.0e-0",
```

File: task_script_utils/parse.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    # --- Reference implementation ---
+    # try:
+    #     return Decimal(value)
+    # except Exception:
+    #     return None
+    raise NotImplementedError("Not implemented yet")
+
+
+def to_boolean(
+    value: str, true_set: Set[str], false_set: Set[str], case_sensitive: bool = False
```
Comment (VladimirShteyn) [inline task_script_utils/parse.py:56]: Should we have a default `true_set` and `false_set`? If so, I recommend

```
true_set: Set[str] = frozenset(('true', 'True', 1, 'yes', 'Yes'))
false_set: Set[str] = frozenset(('false', 'False', '0', 'no', 'No'))
```

File: __tests__/unit/parse_functions/test_to_decimal.py [hunk 6]
```diff
@@ -0,0 +1,353 @@
...
+    """Test Negative decimal with trailing zeros."""
+    # Arrange
+    value = "-56.78000"
+
+    # Act
+    actual = parse.to_decimal(value)
+
+    # Assert
+    assert isinstance(actual, Decimal)
+    assert actual.compare_total(Decimal("-56.78000")) == 0
```
Comment (VladimirShteyn) [inline __tests__/unit/parse_functions/test_to_decimal.py:122]: TIL `compare_total`

**Review by VladimirShteyn — COMMENTED** [review]
Nice work! I just had a few questions.

File: __tests__/unit/parse_functions/test_to_decimal.py (hunk 6 above)
Comment (mlipschultz) [reply __tests__/unit/parse_functions/test_to_decimal.py:122]: Same here!

File: task_script_utils/parse.py (hunk 5 above)
Comment (mlipschultz) [reply task_script_utils/parse.py:56]: For this PR, I think the answer's no.  I think it makes sense to have that discuss as part of this project though.

**Review by igridchyn — APPROVED** [review]
looks awesome! some minor comments

File: __tests__/unit/parse_functions/test_to_int.py (hunk 4 above)
Comment (mlipschultz) [reply __tests__/unit/parse_functions/test_to_int.py:108]: Good idea!  Tests added

File: __tests__/unit/parse_functions/test_to_boolean.py [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+        value, {"True", "1", "yes"}, {"False", "0", "no"}, case_sensitive=case_sensitive
+    )
+
+    # Assert
+    assert actual is False
+
+
+@pytest.mark.parametrize(
+    "value",
+    ["TRUE", "FALSE", "Yes", "yEs", "yeS", "YEs", "YeS", 'yES""YES', "No", "nO", "NO"],
```
[the remark this answers is not in the document]
Comment (mlipschultz) [reply __tests__/unit/parse_functions/test_to_boolean.py:112]: Good catch!

File: __tests__/unit/parse_functions/test_to_boolean.py [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    # Act
+    actual = parse.to_boolean(
+        value, {"True", "1", "yes"}, {"False", "0", "no"}, case_sensitive=case_sensitive
+    )
+
+    # Assert
+    assert actual is True
+
+
+@pytest.mark.parametrize("value", ["False", "0", "no"])
```
[the remark this answers is not in the document]
Comment (mlipschultz) [reply __tests__/unit/parse_functions/test_to_boolean.py:80]: For this test, we just want to know whether a value found in the `true_set` results in `True` being returned.  When it comes time to implement the `to_boolean` function, then we might want to have a comprehensive list.

File: __tests__/unit/parse_functions/test_to_decimal.py [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+
+    # Act
+    actual = parse.to_decimal(value)
+
+    # Assert
+    assert isinstance(actual, Decimal)
+    assert actual.compare_total(Decimal("3.1415")) == 0
+
+
+def test_to_decimal_decimal_representation_of_integer():
```
[the remark this answers is not in the document]
Comment (mlipschultz) [reply __tests__/unit/parse_functions/test_to_decimal.py:266]: It's probably not needed, but it'd be embarrassing if somehow we ended up with that bug.  Test added

File: __tests__/unit/parse_functions/test_to_float.py [hunk 10]
```diff
@@ -0,0 +1,325 @@
+"""Test `to_float()` function in parse.py."""
```
[the remark this answers is not in the document]
Comment (mlipschultz) [reply __tests__/unit/parse_functions/test_to_float.py:1]: I thought we talked about this when deciding on test cases, but I can't find it in the confluence page.  I guess we somehow missed this :disappointed:.  I'll update the confluence page.  Thanks!

File: __tests__/unit/parse_functions/test_to_boolean.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (abclay) [reply __tests__/unit/parse_functions/test_to_boolean.py:80]: The design so far is to not have any comprehensive list of false/true strings.  That is up to the developer to pass into the function.  We cannot assume that we know which values map to False/True in all cases, and the task script dev should know.  It makes the function more clear.

File: __tests__/unit/parse_functions/test_to_boolean.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    # Assert
+    assert actual is None
+
+
+@pytest.mark.parametrize(
+    "value",
+    [
+        "    True",
+        " 1",
+        "\\t   \\t yes",
```
Comment (abclay) [inline __tests__/unit/parse_functions/test_to_boolean.py:135]: This looks like my fault, but I think these `\\t` values should be `\t`

File: __tests__/unit/parse_functions/test_to_float.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mlipschultz) [reply __tests__/unit/parse_functions/test_to_float.py]: Discussion had and decisions made.  Thanks again @igridchyn!

File: __tests__/unit/parse_functions/test_to_boolean.py [hunk 12]
```diff
@@ -109,7 +109,19 @@ def test_to_boolean_case_insensitive_string_is_in_false_set(value):
...
-    ["TRUE", "FALSE", "Yes", "yEs", "yeS", "YEs", "YeS", 'yES""YES', "No", "nO", "NO"],
+    [
+        "TRUE",
+        "FALSE",
+        "Yes",
+        "yEs",
+        "yeS",
+        "YEs",
+        "YeS",
+        'yES", "YES',
```
Comment (mlipschultz) [inline __tests__/unit/parse_functions/test_to_boolean.py:120]: black failed me :angry:

File: __tests__/unit/parse_functions/test_to_boolean.py (hunk 11 above)
Comment (mlipschultz) [reply __tests__/unit/parse_functions/test_to_boolean.py]: Fixed

**Review by abclay — COMMENTED** [review]
I can't approve because I'm the original author, but I approve of these changes anyways.

File: __tests__/unit/parse_functions/test_to_decimal.py [hunk 13]
```diff
@@ -239,93 +239,185 @@ def test_to_decimal_scientific_notation_negative_exponent_greater_than_1():
...
-    value = "128.28       "
+    value = "-512.00"
 
     # Act
     actual = parse.to_decimal(value)
 
     # Assert
     assert isinstance(actual, Decimal)
-    assert actual.compare_total(Decimal("128.28")) == 0
+    assert actual.compare_total(Decimal("-512.00")) == 0
```
Comment (VladimirShteyn) [inline __tests__/unit/parse_functions/test_to_decimal.py:276]: [minor] According to [the compare_total docs](https://docs.python.org/3/library/decimal.html#decimal.Decimal.compare_total) this should be `assert actual.compare_total(Decimal("-512.00")) == Decimal("0")

File: __tests__/unit/parse_functions/test_to_decimal.py [hunk 14]
```diff
@@ -341,6 +433,10 @@ def test_to_decimal_string_has_underscores_decimal():
         "0xbadbeef",
         "0o775",
         "0b100001",
+        "infinity",
+        "INFINITY",
```
Comment (VladimirShteyn) [inline __tests__/unit/parse_functions/test_to_decimal.py:437]: ALL CAPS for when you need a higher cardinality infinity.

**Review by VladimirShteyn — APPROVED** [review]
LGTM. To make the code more readable, might be worth breaking up long strings into multiple lines as suggested [in this reviewer comment](https://github.com/tetrascience/ts-task-script-utils/pull/21/files/35862c5276c7320d42bdd7bd536b4707d86c665d..53675468c0de2c54814503b3d55c682bfa5e98fa#r857863373).

File: __tests__/unit/parse_functions/test_to_decimal.py [hunk 15]
```diff
@@ -0,0 +1,407 @@
...
+
+    # Act
+    actual = parse.to_decimal(value)
+
+    # Assert
+    assert (
+        actual.compare_total(
+            Decimal(
+                "114813069527425452423283320117768198402231770208869520047764273682576626139237031385665948631650626991844596463898746277344711896086305533142593135616665318539129989145312280000688779148240044871428926990063486244781615463646388363947317026040466353970904996558162398808944629605623311649536164221970332681344168908984458505602379484807914058900934776500429002716706625830522008132236281291761267883317206598995396418127021779858404042159853183251540889433902091920554957783589672039160081957216630582755380425583726015528348786419432054508915275783882625175435528800822842770817965453762184851149029376.0"
```
[the remark this answers is not in the document]
Comment (mlipschultz) [reply __tests__/unit/parse_functions/test_to_decimal.py:315]: I was hoping Black would do this for me, but it didn't :angry:

File: task_script_utils/parse.py [hunk 16]
```diff
@@ -19,21 +19,6 @@ def to_int(value: str) -> Optional[int]:
     raise NotImplementedError("Not implemented yet")
 
 
-def to_float(value: str) -> Optional[float]:
```
[the remark this answers is not in the document]
Comment (mlipschultz) [reply task_script_utils/parse.py]: Should we create a `to_float` function that just calls `to_decimal` and convert the output?

[the remark this answers is not in the document]
Comment (abclay) [reply task_script_utils/parse.py]: Would that be part of this effort?  Or would we put the unit tests back and strongly suggest the implementer decide to use `to_decimal` as the parser within `to_float`?

File: __tests__/unit/parse_functions/test_to_decimal.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (abclay) [reply __tests__/unit/parse_functions/test_to_decimal.py]: for really long ints, we could do something like 

```python
expected = int(
    "123456790"
    "123456790"
    "123456790"
    "123456790"
)
assert expected == actual
```
with the understanding that we trust the `int()` built-in.  Or we could just leave them as-is and mark these files telling pylint to ignore long lines in them.

File: task_script_utils/parse.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (abclay) [reply task_script_utils/parse.py]: +1 to bringing back `to_float` for right now.  We can base its implementation on `to_decimal` and deprecate it later if we want.  Folks should be able to switch over to `to_decimal` easily since it's the same implementation and behaves the same except for the output type.

[the remark this answers is not in the document]
Comment (mlipschultz) [reply task_script_utils/parse.py]: Brought back and added tests for nan and inf

**Review by abclay — COMMENTED** [review]
LGTM
